File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2019/Y19D09.kt
package org.gristle.pdxpuzzles.advent.y2019

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.advent.y2019.Intcode.FintCode


class Y19D09(private val input: String) : Day {

    override fun part1() = FintCode(input).run(listOf(1L))[0]

    override fun part2() = FintCode(input).run(listOf(2L))[0]
}

fun main() = Day.runDay(Y19D09::class)

//    Class creation: 19ms
//    Part 1: 2870072642 (2ms)
//    Part 2: 58534 (84ms)
//    Total time: 105ms

File: rust/advent/src/bin/2019/y19d09.rs
use advent::utilities::get_input::get_input;
use advent::utilities::intcode::{IntCode, State};
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = i64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(19, 9).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(input: Input, initial: i64) -> Output {
    let mut comp = IntCode::from(input);
    comp.input(initial);
    let State::Output(value) = comp.run() else { unreachable!() };
    value
}

fn part1(input: Input) -> Output {
    solve(input, 1)
}

fn part2(input: Input) -> Output {
    solve(input, 2)
}

#[test]
fn default() {
    let input = get_input(19, 9).unwrap();
    assert_eq!(2870072642, part1(&input));
    assert_eq!(58534, part2(&input));
}

// Input parsed (17μs)
// 1. 2870072642 (41μs)
// 2. 58534 (2ms)
// Total: 2ms